### PR TITLE
[docs] Create reviewing.md

### PR DESCRIPTION
**Pull request by SSE4**
this is mostly a follow up to the discussion started at #3390

from time to time some contributors feel frustration that their pull requests take weeks and months to be reviewed and approved, this happens because reviewers request changes (or do code suggestions) for the minor code-style or formatting things, such as:

- remove trailing white-spaces
- change double quotes to single quotes
- re-order methods

(the list is incomplete, just a few common patterns I have on my mind right now).

while these code style changes can make code easier to read for some people, they don't bring much value for consumers on their own (e.g. they don't add support for more configurations, generators, versions, etc.). but each time such minor code-style changes are applied, they invalidate all existing reviews, making PRs longer to process. also, they introduce more line changes in patch, so it might require more attention to check if there are actual code changes in certain lines, or just formatting/styling.

therefore, I'd like to start the discussion, if we want to formalize and document these small things, and decide, if we want to enforce them (e.g. via `editorconfig`, conan hooks, linters or 3rd-party tools like [black](https://github.com/psf/black) or [flake8](https://gitlab.com/pycqa/flake8)).

at very least, it would be nice to have a consensus, if we ever want to formalize these things, if we want to enforce such policies in PRs, and if we want to have some automated checks for them.

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: docs/reviewing.md
Comment (madebr) [inline docs/reviewing.md:11]: This is hot food for debate.
See e.g. this twitter message: https://twitter.com/raymondh/status/1259209765072154624

I would change it to the following:
```
Don't mix single quotes (`'`) and double quotes (`"`) in python code (`conanfile.py`, `test_package/conanfile.py`).
Be consistent.
```

**Comment by prince-chrismc** [comment]
The title of the doc is a little misleading, it's just some syntaxing guidelines, are we planning on adding more

**Comment by ericLemanissier** [comment]
I don't know how it could be formalized in a document, but I think that it's fine if cosmetic changes are requested on PR which has not yet been approved by any one yet. OTOH, after the PR has been approved at some point (even if this approval is outdated because another change was made afterwards), reviewers should highly refrain from requesting cosmetic changes, or changes which make no difference to consumers of the package. This kind of request only brings frustration  and delay to the situation, and could dissuade contributors to propose other PRs. In this case, it is better if the changes are made in a follow-up PR, and I guess that sometimes, they won't seem as necessary once the original PR is merged.
(Of course, I'm one of the offenders to this "guideline" or whatever this is :sweat:)

File: docs/reviewing.md
Comment (SSE4) [reply docs/reviewing.md:11]: yes, I agree.

**Comment by SSE4** [comment]
> As a small aside, the title of the doc is a little misleading as it stands since it's just around syntaxing guidelines. Are we planning on adding more?

yes, if you have some common pattern on mind, feel free to add.

> 
> Should this be apart of the contributing guidelines as well?

I don't know yet. first step is to collect feedback. it's very possible we won't be forcing any of these guidelines at all, providing a complete freedom for syntax and styling. we need to know what contributors and reviewers think about it, do they find these things listed useful, annoying, frustrating, etc.

**Comment by SSE4** [comment]
to clarify, the frustrating scenario is:
1. contributor submits a new PR
2. PR passes CI
3. reviewer A submits an approval
4. reviewer B requests syntax/style/formatting changes
5. contributor accepts changes requested
6. approval from A is invalidated
7. steps 3-6 repeat several times

the ideal scenario would be:
1. contributor submits a new PR
2. CI reports syntax/style/formatting violation
3. contributor fixes CI complains
4. PR passes CI
5. reviewer A submits an approval
6. reviewer B submits an approval

in the second case, there are no approvals invalidated due to the syntax/style/formatting changes, which should cause no additional review delays.

of course the situation could be solved without automation as well, if all reviewers agree to be faithful - e.g. never request syntax/style/formatting changes that might potentially invalidate existing reviews.

**Review by danimtb — APPROVED** [review]
I agree we should document this kind of policies and this is a good start

File: docs/reviewing.md
Comment (jgsogo) [inline docs/reviewing.md:15]: I think there is a logical order for Conan methods, right? At least this is how I think about Conan recipes, it happens that for many of them the lexicographic ordering is equal to the logical one, but I prefer to use something meaningful.... otherwise, `requirements()` will be moved at the bottom of the `conanfile.py` (and I prefer to have it closer to the class attributes).

Comment (Croydon) [reply docs/reviewing.md:15]: @SpaceIm has proposed to use the order in which the methods are executed by Conan and I think that makes a lot of sense

Comment (jgsogo) [reply docs/reviewing.md:15]: I would expect them to match the order they are documented, which probably should match the order they are executed (now doc has no order, just in case: https://docs.conan.io/en/latest/reference/conanfile/methods.html). Not sure if it is exactly the order they are executed (for example, and this is my very personal opinion, I prefer `package_id()` at the bottom, but it is usually executed when building the graph to search for available binaries).

Comment (SSE4) [reply docs/reviewing.md:15]: is there strict execution order? I think some methods could be called several times at different stages

Comment (prince-chrismc) [reply docs/reviewing.md:15]: 💭 I assume CCI only calls `create` to make the binaries.... could that be the order?

Comment (SSE4) [reply docs/reviewing.md:15]: not really. it first computes package ids to build and understand how many agents and of what type (win/linux/mac) to spawn. then it probably calls `create`. and afterwards, `upload` at some later stage.

Comment (madebr) [reply docs/reviewing.md:15]: alpha-numeric order does not make sense.

Methods that work together should be grouped together.
Putting them in order that conan calls them makes total sense.

I also try to order the attributes at the top of a Conan file class:
- put name, description, license, homepage, topics at the top as these are user facing
- then url
- then options + default_options
- settings
- then all options that are build specific: options that must be set to make the package build.
Such as generators, no_copy_source, exports_sources, ...

The ordering of top attributes might be nitpicking though

Comment (madebr) [reply docs/reviewing.md:15]: > not really. it first computes package ids to build and understand how many agents and of what type (win/linux/mac) to spawn. then it probably calls `create`. and afterwards, `upload` at some later stage.

I wouldn't take the conan-package-tools as a reference.
It does not fully initialize the recipe.
(early revisions of my boost pr at https://github.com/conan-io/conan-center-index/pull/2097 failed to build on conan-package-tools because of this)

Comment (SSE4) [reply docs/reviewing.md:15]: I don't mind doing an experiment (logging the order of methods executed on Conan Center Index CI) and documenting this order here.

Comment (SSE4) [reply docs/reviewing.md:15]: did some experiment in #3636. the current order of methods on CI:
- init
- set_name
- set_version
- export
- export_sources
- config_options
- configure
- requirements
- package_id
- build_id
- build_requirements
- system_requirements
- source
- imports
- build
- package
- package_info
never called on CI: 
- deploy
- test

should we document this order as preferred for recipes?
I don't mind, but keep in mind:
- conan implementation may change, and order of execution is subject to change as well (as long as it doesn't break, I think some methods could be safely swapped)
- CI implementation may change as well to call different commands in different order

Comment (SSE4) [reply docs/reviewing.md:15]: btw, another question, while we can have order of documented conanfile methods defined, what should we do with the following:
- private recipe methods (starting with `__`)
- so called `magic` methods (such as `__new__`, `__init__`, `__del__` and so on)
- protected recipe methods (starting with `_`)
- other (non-documented) public methods

in general, do we expect them to appear before/after documented `ConanFile` methods such as `source`?

Comment (Croydon) [reply docs/reviewing.md]: I don't think we should over-specify, having a default order for the Conan methods should cover most things already

Regarding custom methods as an example, in CMake based recipes we almost always have a `_cmake_configure()` method, where IMO it makes sense to have this directly above `build()`. Other private methods might not be that common and might only do little things like string manipulations/mappings. It might not always be that clear where to put them as in the case of `_cmake_configure()`. I would leave that to whoever is writing the recipe

Regarding enforcement, unfortunately I don't think we can add for this a nice config file like `.gitattributes` or `.editorconfig` which "just works" for most contributors, but if core contributors use this order + when we try to use this order in documentation and templates, then we will get it more standardized than before without annoying people too much for something non-functional.

Comment (madebr) [reply docs/reviewing.md]: What @Croydon says. Let's recommend, not require or enforce.

**Comment by prince-chrismc** [comment]
One item that came to mind was the `test_package`

Review comments I see often:
- keeping it minimal
- c and/or cpp
- does it actually test linkage to all deps
- cmake conan targets VS find_package
- conanfile structure

Which of the items above are *required* and which are *prefered* (ie. belong in this doc)

I think this is a separate topic but the structure of test_package. it's very much boilerplated for every recipe however on occasion one crops up that quite far... but Croydon comments would apply to it as well

> when we try to use this [...] in documentation and templates, then we will get it more standardized than before

**Review by uilianries — APPROVED** [review]
This current and initial version looks good. It's out first step and we can improve a lot on next PRs.

**Review by danimtb — APPROVED** [review]

File: docs/reviewing.md
Comment (mathbunnyru) [inline docs/reviewing.md:37]:
```suggestion
the order above resembles the execution order of methods on CI. therefore, for instance, `build` is always executed before `package` method, so `build` should appear before the `package` in `conanfile.py`.

## Simultaneous conflicting methods

Try to avoid mixing the `requires` field and `requirements()` method in the same `conanfile.py`. It's better to put requirements close to each other.
The same thing goes about `build_requires` and `build_ requirements()`.
```

File: docs/reviewing.md
Comment (mathbunnyru) [inline docs/reviewing.md:11]:
```suggestion
If possible, try to avoid mixing single quotes (`'`) and double quotes (`"`) in python code (`conanfile.py`, `test_package/conanfile.py`). Consistency is preferred.
```

File: docs/reviewing.md
Comment (mathbunnyru) [inline docs/reviewing.md:15]: Where should we put `validate`?

**Review by mathbunnyru — CHANGES_REQUESTED** [review]

File: docs/reviewing.md
Comment (mathbunnyru) [reply docs/reviewing.md:37]: @uilianries created a hook, so this change is not needed.
https://github.com/conan-io/hooks/pull/261

**Review by prince-chrismc — APPROVED** [review]
🤗

**Pull request merged**
